tests/pinmaplat.cpp

```diff
@@ -0,0 +1,207 @@
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <getopt.h>
+#include <memory.h>
+#include <stdio.h>
+#include <math.h>
+#include <iostream>
+#include <iomanip>
+#include <cuda.h>
+#include <cuda_runtime_api.h>
+
+using namespace std;
+
+#include "gdrapi.h"
+#include "common.hpp"
+
+using namespace gdrcopy::test;
+
+// manually tuned...
+int num_iters = 100;
+
+int main(int argc, char *argv[])
+{
+    size_t _size = (size_t)1 << 24;
+    // minimum pinning size is a GPU page size
+    size_t pin_request_size = GPU_PAGE_SIZE;
+    int dev_id = 0;
+    struct timespec beg, end;
+    double pin_lat_us;
+    double map_lat_us;
+    double unpin_lat_us;
+    double unmap_lat_us;
+    double inf_lat_us;
+
+    while(1) {        
+        int c;
+        c = getopt(argc, argv, "s:d:n:h");
+        if (c == -1)
+            break;
+
+        switch (c) {
+            case 's':
+                _size = strtol(optarg, NULL, 0);
+                break;
+            case 'd':
+                dev_id = strtol(optarg, NULL, 0);
+                break;
+            case 'n':
+                num_iters = strtol(optarg, NULL, 0);
+                break;
+            case 'h':
+                printf("syntax: %s -s <buf size> -d <gpu dev id> -n <num iters> -h[help]\n", argv[0]);
+                exit(EXIT_FAILURE);
+                break;
+            default:
+                printf("ERROR: invalid option\n");
+                exit(EXIT_FAILURE);
+        }
+    }
+    
+    size_t size = (_size + GPU_PAGE_SIZE - 1) & GPU_PAGE_MASK;
+
+    ASSERTDRV(cuInit(0));
+
+    int n_devices = 0;
+    ASSERTDRV(cuDeviceGetCount(&n_devices));
+
+    CUdevice dev;
+    for (int n=0; n<n_devices; ++n) {
+        
+        char dev_name[256];
+        int dev_pci_domain_id;
+        int dev_pci_bus_id;
+        int dev_pci_device_id;
+
+        ASSERTDRV(cuDeviceGet(&dev, n));
+        ASSERTDRV(cuDeviceGetName(dev_name, sizeof(dev_name) / sizeof(dev_name[0]), dev));
+        ASSERTDRV(cuDeviceGetAttribute(&dev_pci_domain_id, CU_DEVICE_ATTRIBUTE_PCI_DOMAIN_ID, dev));
+        ASSERTDRV(cuDeviceGetAttribute(&dev_pci_bus_id, CU_DEVICE_ATTRIBUTE_PCI_BUS_ID, dev));
+        ASSERTDRV(cuDeviceGetAttribute(&dev_pci_device_id, CU_DEVICE_ATTRIBUTE_PCI_DEVICE_ID, dev));
+
+        cout  << "GPU id:" << n << "; name: " << dev_name 
+              << "; Bus id: "
+              << std::hex 
+              << std::setfill('0') << std::setw(4) << dev_pci_domain_id
+              << ":" << std::setfill('0') << std::setw(2) << dev_pci_bus_id
+              << ":" << std::setfill('0') << std::setw(2) << dev_pci_device_id
+              << std::dec
+              << endl;
+    }
+    cout << "selecting device " << dev_id << endl;
+    ASSERTDRV(cuDeviceGet(&dev, dev_id));
+
+    CUcontext dev_ctx;
+    ASSERTDRV(cuDevicePrimaryCtxRetain(&dev_ctx, dev));
+    ASSERTDRV(cuCtxSetCurrent(dev_ctx));
+
+    CUdeviceptr d_A;
+    ASSERTDRV(cuMemAlloc(&d_A, size));
```

Comment:
cuMemAlloc doesn’t guarantee alignment. I suggest that you use gpuMemAlloc instead.